File: SEM-2/DAA/ISE/movie_ticket_booking_backend/controllers/showController.js
// controllers/showController.js
const Show = require('../models/Show');

// Retrieve all shows, populated with movie and theatre details
exports.getAllShows = async (req, res) => {
  try {
    const shows = await Show.find().populate('movie').populate('theatre');
    res.json(shows);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
};

// Create a new show
exports.createShow = async (req, res) => {
  try {
    const show = new Show(req.body);
    await show.save();
    res.status(201).json(show);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
};

// Book seats by removing the booked seats from availableSeats
exports.bookSeats = async (req, res) => {
  try {
    const show = await Show.findById(req.params.id);
    if (!show) return res.status(404).json({ message: 'Show not found' });

    const { seatNumbers } = req.body;
    // Check if all requested seats are available
    const available = seatNumbers.every(seat => show.availableSeats.includes(seat));
    if (!available) {
      return res.status(400).json({ message: 'Some seats are not available' });
    }

    // Remove booked seats
    show.availableSeats = show.availableSeats.filter(seat => !seatNumbers.includes(seat));
    await show.save();
    res.json({ message: 'Seats booked successfully', show });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
};

File: SEM-2/DAA/ISE/movie_ticket_booking_application/lib/controllers/show_controller.dart
import '../models/show.dart';
import '../utils/api_service.dart';

class ShowController {
  final ApiService _apiService = ApiService();

  Future<List<Show>> getAllShows() async {
    return await _apiService.fetchShows();
  }

  Future<List<Show>> getShowsByMovie(String movieId) async {
    return await _apiService.fetchShowsByMovie(movieId);
  }

  Future<Show> bookSeats(String showId, List<int> seatNumbers) async {
    return await _apiService.bookSeats(showId, seatNumbers);
  }
}